### ISSUE TITLE
Windows CMakeLists.txt BINARY_NAME issue

### ISSUE THREAD

**Issue by kirill-21** [flutter/flutter#98994]
Steps to reproduce:
- Create new Flutter project on Windows
- Go to windows/CMakeLists.txt and replace
```
set(BINARY_NAME "tesee1")
```
with
```
set(APPLICATION_TYPE 0)
add_compile_definitions(IS_EXLOADER=${APPLICATION_TYPE})

if(APPLICATION_TYPE)
  set(BINARY_NAME "tesee1")
else()
  set(BINARY_NAME "example2")
endif()
```
- Start application(clean build folder if you've compiled app before)

For some reason flutter creates .exe file with filename example2 but tries to start tesee1
![image](https://user-images.githubusercontent.com/63228361/155365272-0c0216fe-ae74-4f2e-bcd2-1904d2642822.png)

Flutter version: 2.10.2 from stable channel

**Comment by kirill-21** [comment]
> BINARY_NAME

![image](https://user-images.githubusercontent.com/63228361/155409572-6f7f21a0-d868-4d04-983a-9cc8178072ef.png)
Yet BINARY_NAME is the recommended way of changing the name of the generated binary.

What i'm trying to achieve is one code base with two whole different UI implementations. Just switch one variable in CMake and app will be rendered differently

**Comment by stuartmorgan** [comment]
> Yet BINARY_NAME is the recommended way of changing the name of the generated binary.

Yes, it is. Specifically, by changing the value (the comment says "Change this line" not "Add arbitrary logic around this line").

The Flutter tool doesn't support all possible changes to the native build files, on any platform. If you change things that Flutter relies on, the tool can't drive it correctly any more.

> What i'm trying to achieve is one code base with two whole different UI implementations. Just switch one variable in CMake and app will be rendered differently

It sounds like you want flavor support for Windows; updating description accordingly.

**Comment by kirill-21** [comment]
> flavor support for Windows

I do not know what is flavor, just to avoid unnecessary renames every build i tried to implement dynamic executable name

**Comment by insertjokehere** [comment]
This is something I'm very interested in as well. For a first pass, would it make sense to modify the generated CMake config here:

https://github.com/flutter/flutter/blob/6dae269881b426c30ff214a1c719671c61dfd102/packages/flutter_tools/lib/src/cmake.dart#L100-L125

To set a `FLUTTER_BUILD_FLAVOR` variable (using the value passed in by the `BuildInfo` object) that could be referenced in the other CMake files to do whatever conditional logic you need for your different flavors. In our case this is picking a different icon, linking in a different resource file, and copying different public keys into the install directory.

You'd also need to modify 

https://github.com/flutter/flutter/blob/6dae269881b426c30ff214a1c719671c61dfd102/packages/flutter_tools/lib/src/commands/build_windows.dart#L21-L26

To call `usesFlavorOption()` similar to how `build_appbundle.dart` does it.

I'm happy to spend a bit of time putting together a PR for this, but before I do can someone from the desktop team speak to if a design like this would be acceptable?

**Comment by kirill-21** [comment]
> This is something I'm very interested in as well. For a first pass, would it make sense to modify the generated CMake config here:
> 
> https://github.com/flutter/flutter/blob/6dae269881b426c30ff214a1c719671c61dfd102/packages/flutter_tools/lib/src/cmake.dart#L100-L125
> 
> To set a `FLUTTER_BUILD_FLAVOR` variable (using the value passed in by the `BuildInfo` object) that could be referenced in the other CMake files to do whatever conditional logic you need for your different flavors. In our case this is picking a different icon, linking in a different resource file, and copying different public keys into the install directory.
> 
> You'd also need to modify
> 
> https://github.com/flutter/flutter/blob/6dae269881b426c30ff214a1c719671c61dfd102/packages/flutter_tools/lib/src/commands/build_windows.dart#L21-L26
> 
> To call `usesFlavorOption()` similar to how `build_appbundle.dart` does it.
> 
> I'm happy to spend a bit of time putting together a PR for this, but before I do can someone from the desktop team speak to if a design like this would be acceptable?

I think it all should be controlled in pubspec.yaml and be available for all platforms, otherwise it will not be added :c

**Comment by andrewkolos** [comment]
> What i'm trying to achieve is one code base with two whole different UI implementations. Just switch one variable in CMake and app will be rendered differently

Would an environment variable be sufficient for this? You could use `--dart-define=MY_ENV_VAR=value` in your `flutter build`/`flutter run` scripts and get the value at run time using `const String.fromEnvironment('MY_ENV_VAR')`.

**Comment by kirill-21** [comment]
> > What i'm trying to achieve is one code base with two whole different UI implementations. Just switch one variable in CMake and app will be rendered differently
> 
> Would an environment variable be sufficient for this? You could use `--dart-define=MY_ENV_VAR=value` in your `flutter build`/`flutter run` scripts and get the value at run time using `const String.fromEnvironment('MY_ENV_VAR')`.

This won't affect app's icon, metadata, name

**Comment by insertjokehere** [comment]
> > > What i'm trying to achieve is one code base with two whole different UI implementations. Just switch one variable in CMake and app will be rendered differently
> > 
> > 
> > Would an environment variable be sufficient for this? You could use `--dart-define=MY_ENV_VAR=value` in your `flutter build`/`flutter run` scripts and get the value at run time using `const String.fromEnvironment('MY_ENV_VAR')`.
> 
> This won't affect app's icon, metadata, name

We also use some custom CMake to conditionally include some extra files in the build output - Firebase keys, and a public key for verifying upgrades. Doing this through flavors makes it easier to only include the files we need for specific versions. Doing this through assets results in all versions of the files being bundled in the build output unless we manually modify the asset files before doing the build which seems like a bit of a hack. I just want Windows to have feature parity with the other platforms.

**Comment by talamaska** [comment]
The flavours support is lacking for Linux as well. Seems like only iOS, MacOs and Android have this feature. Having this flagged as P3 is not sending a good message to all those who waited patiently for the Windows support all these years.

**Comment by Stumblinbear** [comment]
With conditionally bundling of assets based on flavors, having support for flavors in desktop environments has more uses and should (imo) be a bit of a higher priority

**Comment by 12people** [comment]
I created a [corresponding ticket for Linux](https://github.com/flutter/flutter/issues/162373), so that all desktop platforms can have equal flavor support.

**Comment by kjxbyz** [comment]
This feature is very important. I need to package the same code into a free version and a paid version. Then upload it to the Microsoft App Store

**Comment by lucavenir** [comment]
Years later...

Is there any plan towards this issue, and - more generally - offering flavors to _all_ platforms, without hiccups?

This implies:
* addressing this issue
* re-opening / addressing the [web issue](https://github.com/flutter/flutter/issues/59388) (why is it closed..?)
* addressing [the linux issue](https://github.com/flutter/flutter/issues/162373)

Even [the docs](https://docs.flutter.dev/deployment/flavors) about this topic are kind weird. The path is `docs.flutter.dev/deployment/flavors`, but the page title clearly states `Set up Flavors for Android`. You can then jump to [its sibling page](https://docs.flutter.dev/deployment/flavors-ios), `docs.flutter.dev/deployment/flavors-ios`. Don't get me wrong those docs are alright. They just reflect the state Flutter is in, when it comes to flavoring.

I don't think `P3` is the right priority for this issue.
